libfeed/feed-knowledge-app-card-store.c

```diff
@@ -0,0 +1,297 @@
+/* Copyright 2018 Endless Mobile, Inc.
+ *
+ * eos-discovery-feed is free software: you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation, either version 2.1 of the
+ * License, or (at your option) any later version.
+ *
+ * eos-discovery-feed is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with eos-discovery-feed.  If not, see
+ * <http://www.gnu.org/licenses/>.
+ */
+
+#include "feed-knowledge-app-card-store.h"
+#include "feed-sizes.h"
+#include "feed-util.h"
+
+struct _EosDiscoveryFeedKnowledgeAppCardStore
+{
+  GObject parent_instance;
+};
+
+typedef struct _EosDiscoveryFeedKnowledgeAppCardStorePrivate
+{
+  gchar                               *title;
+  gchar                               *uri;
+  gchar                               *synopsis;
+  gchar                               *bus_name;
+  gchar                               *knowledge_search_object_path;
+  gchar                               *knowledge_app_id;
+  GInputStream                        *thumbnail;
+  EosDiscoveryFeedCardLayoutDirection  layout_direction;
+  guint                                thumbnail_size;
+} EosDiscoveryFeedKnowledgeAppCardStorePrivate;
+
+G_DEFINE_TYPE_WITH_PRIVATE (EosDiscoveryFeedKnowledgeAppCardStore,
+                            eos_discovery_feed_knowledge_app_card_store,
+                            EOS_DISCOVERY_FEED_TYPE_APP_CARD_STORE)
+
+enum {
+  PROP_0,
+  PROP_TITLE,
+  PROP_URI,
+  PROP_SYNOPSIS,
+  PROP_THUMBNAIL,
+  PROP_BUS_NAME,
+  PROP_KNOWLEDGE_SEARCH_OBJECT_PATH,
+  PROP_KNOWLEDGE_APP_ID,
+  PROP_LAYOUT_DIRECTION,
+  PROP_THUMBNAIL_SIZE,
+  NPROPS
+};
+
+static GParamSpec *eos_discovery_feed_knowledge_app_card_store_props [NPROPS] = { NULL, };
+
+static GObject *
+eos_discovery_feed_knowledge_app_card_store_constructor (GType                  type,
+                                                         guint                  n_construct_params,
+                                                         GObjectConstructParam *construct_params)
+{
+  g_auto(GValue) value = G_VALUE_INIT;
+
+  maybe_append_card_type_construct_prop (construct_params,
+                                         n_construct_params,
+                                         EOS_DISCOVERY_FEED_CARD_STORE_TYPE_ARTICLE_CARD);
+
+  return G_OBJECT_CLASS (eos_discovery_feed_knowledge_app_card_store_parent_class)->constructor (type,
+                                                                                                 n_construct_params,
+                                                                                                 construct_params);
+}
```

Comment:
Note: Overriding the constructor in order to change the default value without also overriding the prop implementation is a little clunky - I am not sure if it is possible to use `g_object_class_install_property` to *just* change the defaults without also overriding the implementation (it appears as though a new prop id has to be assigned for it).

Docs for `g_object_class_install_property` say that: "Note that it is possible to redefine a property in a derived class, by installing a property with the same name. This can be useful at times, e.g. to change the range of allowed values or the default value."

https://developer.gnome.org/gobject/stable/gobject-The-Base-Object-Type.html#g-object-class-install-property

Reply:
Yeah, please let's find a way not to override `constructor` for the card type property.  I am not even sure it has to be a writable property?